### PR TITLE
FP-FIX(architecture): eliminate false positives across architecture/correctness/design rules

### PR DESCRIPTION
**Pull request by aidenybai**
## FP-FIX — Architecture / correctness / design rules

Part of the false-positive hardening stack. Removes false positives across the **architecture, correctness, and design** families. Based on #988 (foundation) — merge that first.

### Highlights
- **`no-many-boolean-props`** — only counts genuine on/off props: boolean-prefixed names invoked as callbacks (`props.showMenu()`) or wired as JSX event handlers (`onClick={props.showMenu}`) are excluded for both the destructured and `props`-object shapes; also requires real render output before treating a param as component props.
- **`no-render-in-render`** — `rootsInProps` traces render props destructured from nested prop bags, so legitimate render-prop calls aren't flagged.
- **`no-nested-component-definition`** — only a PascalCase binding actually rendered as JSX (`<Name/>`) is flagged; one exclusively invoked as `Name()` is inlined, not a remounting child.
- **`react-compiler-no-manual-memoization` / `react-compiler-destructure-method` / `prefer-module-scope-static-value`** and the `correctness/` (invalid-paragraph-child, nested-interactive, polymorphic-children, uncontrolled-input, prevent-default, svg-precision) + `design/` (outline-none, gray-on-colored, transition rules, side-tab-border) rules tightened.

### Bugbot follow-ups
- **`no-many-boolean-props` handler exclusion incomplete** — FIXED: the `props`-object path now also excludes names wired as `onClick={props.x}`, matching the destructured path.
- **`no-nested-component-definition` file-wide JSX match** — declined (consistent with the earlier call on the combined PR): making it precise requires scope/binding resolution to prove a `<Name/>` usage resolves to *this* nested definition rather than an import or a same-named sibling. The residual FP is rare; tracked rather than shipping a half-measure. (See thread.)

### Test plan
- [x] architecture / correctness / design unit + regression suites green
- [x] typecheck clean
- [ ] Full CI after foundation merges + retarget to `main`

## RDE eval results

Validated all 18 rules in this PR against the OSS corpus with the `react-doctor-evals` harness (**local** runner — cloud reports a misleading `0` for oxlint AST rules).

| Check                    | Result                                                              |
| ------------------------ | ------------------------------------------------------------------- |
| Repos scanned (distinct) | `64`                                                                |
| RootDir scans            | `500` (495 reported, 5 failed)                                      |
| Manifest                 | `repos.json` (8,423 entries)                                        |
| Target diagnostics       | `~8,490`                                                            |
| False positives found    | `1` class — `no-uncontrolled-input` on `disabled` inputs (3×)       |
| Fix                      | exempt `disabled` inputs + regression tests (landed in this PR)     |

Per-rule (every hit sampled/inspected — all true positives by contract except the one FP class, now fixed):

| Rule                                   | Hits | Notes                                                              |
| -------------------------------------- | ---- | ----------------------------------------------------------------- |
| rendering-svg-precision                | 5379 | over-precise machine-exported SVG; ≥2-distinct-token gate verified |
| react-compiler-no-manual-memoization   | 1463 | only fires where React Compiler is configured (verified)          |
| no-render-in-render                    | 692  | local render helpers, not render props (carve-out verified)       |
| prefer-module-scope-static-value       | 491  | static config/arrays; impure carve-out correct                    |
| no-many-boolean-props                  | 176  | genuine boolean props                                             |
| no-nested-component-definition         | 103  | real nested components rendered as JSX                            |
| no-prevent-default                     | 46   | forms without `action=` / non-navigating anchors                  |
| no-outline-none                        | 44   | `outline:none` without a focus ring                               |
| no-polymorphic-children                | 35   | `typeof children` resolving to props                              |
| html-no-invalid-paragraph-child        | 30   | invalid `<p>` descendants                                         |
| no-render-prop-children                | 13   | ≥3 real render props (no `*Props` bags / literals)                |
| no-layout-transition-inline            | 6    | animating width/height                                            |
| html-no-nested-interactive             | 5    | nested interactive elements                                       |
| no-long-transition-duration            | 4    | >1000 ms                                                          |
| **no-uncontrolled-input**              | 2    | TPs (value+defaultValue; undefined-state) after the `disabled` fix |
| no-gray-on-colored-background / no-side-tab-border / react-compiler-destructure-method | 0 | no hits in corpus |

The lone false positive — `<input value={…} disabled />` (supabase's read-only email field, ported 3 ways) — is fixed in this PR by `fix(no-uncontrolled-input): exempt disabled inputs`, backed by a new regression suite. React suppresses its missing-`onChange` warning for `disabled` fields exactly like `readOnly`.

---

> [!NOTE]
> **Low Risk**
> Changes are precision-focused lint heuristics with extensive regression tests; risk is mainly missed true positives on edge cases, not runtime or security impact.
> 
> **Overview**
> This PR **reduces false positives** across ~15 oxlint-plugin-react-doctor rules in architecture, correctness, and design, while keeping the underlying smells. Shared helpers (`isComponentFunction`, `isComponentParameterSymbol`, scope-aware props tracing) and test harness fixes (`:exit` visitors) underpin several rule changes.
> 
> **Architecture** rules now distinguish real components from factories and callbacks from boolean flags: render-output gates, callback/event-handler exclusions, deferred reporting for nested components only when actually rendered (JSX or `component={…}`), render-prop carve-outs, impure-global / scalar-lookup abstentions for hoisting advice, and React Compiler carve-outs for `useSearchParams` and custom `memo` comparators.
> 
> **Correctness** rules stop treating JSX passed as non-`children` props as DOM ancestry, resolve `children` / props via scope, relax `preventDefault` / uncontrolled-input checks for progressive enhancement and React-aligned cases, and add an SVG precision materiality threshold plus test/docs paths.
> 
> **Design** rules tighten Tailwind variant pairing for gray-on-color, exact layout-transition tokens, infinite/`aria-hidden` animation exemptions, stricter focus-ring detection for `outline: none`, and achromatic arbitrary border colors for side-tab borders.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7ddc33b10232d2c5baab2e85f7d679455dc49ca4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-many-boolean-props.ts [hunk 1]
```diff
@@ -58,18 +90,26 @@ export const noManyBooleanProps = defineRule({
...
+      // actual render output before treating the param as component props.
+      if (!functionContainsReactRenderOutput(functionNode, context.scopes)) return;
       if (isNodeOfType(param, "ObjectPattern")) {
+        const callbackUsedNames = collectCallbackUsedNames(body);
         const booleanLikePropNames: string[] = [];
         for (const property of param.properties ?? []) {
           if (!isNodeOfType(property, "Property")) continue;
           const keyName = isNodeOfType(property.key, "Identifier") ? property.key.name : null;
           if (!keyName) continue;
+          if (callbackUsedNames.has(keyName)) continue;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-many-boolean-props.ts]: Fixed in 5227343c. The `props`-object path now also excludes boolean-prefixed members wired as JSX event handlers (`onClick={props.showMenu}`), matching the destructured-param exclusion — not just invoked callbacks (`props.showMenu()`). Regression test added.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-nested-component-definition.ts [hunk 2]
```diff
@@ -35,16 +56,26 @@ export const noNestedComponentDefinition = defineRule({
...
         if (isComponentAssignment(node)) componentStack.pop();
       },
+      "Program:exit"() {
+        for (const candidate of candidates) {
+          if (!renderedJsxNames.has(candidate.name)) continue;
+          context.report({
+            node: candidate.reportNode,
+            message: `Your users lose all state in "${candidate.name}" on every render because it's defined inside "${candidate.enclosingName}", so move it out to the top of the file.`,
+          });
+        }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-nested-component-definition.ts:107]: Declining (consistent with the same call earlier on the combined PR). The file-wide JSX-name match is a deliberate simplification. Making it precise — proving that a given `<Name/>` usage resolves to *this* nested definition's binding rather than an import or a sibling component's same-named definition — requires scope/binding resolution this rule doesn't do. The residual FP (a nested helper used only via `Name()` while another component in the same module renders a same-named `<Name/>`) is rare and was judged not worth the binding-resolution machinery for v1. Tracking separately rather than shipping a half-measure that could mis-resolve.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-render-in-render.ts [hunk 3]
```diff
@@ -1,8 +1,90 @@
...
+// the idiomatic React render-prop pattern, not inline component
+// construction.
+const isRenderPropReceiver = (object: EsTreeNode): boolean => {
+  if (isNodeOfType(object, "Identifier")) return object.name === "props";
+  return (
+    isNodeOfType(object, "MemberExpression") &&
+    isNodeOfType(object.property, "Identifier") &&
+    object.property.name === "props"
+  );
+};
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-render-in-render.ts]: Fixed in b98569a5. `isRenderPropReceiver` now only exempts the bare `props` identifier or `this.props` — an unrelated `someObject.props.renderX()` no longer hides a real inline render call. Regression test added.

**Comment by rayhanadev** [comment]
🧹 **Deslop + changeset pass** (commit `c421da60`)

Ran a heavy deslop over this PR's FP-hardening changes. The diff is genuinely clean — most candidates resolved to *"leave as-is: matches the neighboring rule / load-bearing"* (the `no-nested-component-definition` deferred-report buffer is required for forward JSX refs; the `html-no-*` "why" comments match their siblings; the `no-long-transition-duration` split-brain detection is complementary). Applied only the real, behavior-preserving wins:

- **Consolidated a duplicate** — the `on*` event-handler-attribute predicate appeared twice in `no-many-boolean-props`, and the codebase already names this concept inline in `tanstack-start`. Extracted it into a shared `isEventHandlerAttribute` type-guard util (`utils/is-event-handler-attribute.ts`) and repointed both sites.
- **Fixed a misplaced test** — the `memo(_, areEqual)` `expectDiagnosticCount(…, 0)` case was sitting in the *fail-cases* `describe`; moved it to *pass-cases*.
- **Added the missing changeset** (`oxlint-plugin-react-doctor` patch) documenting this PR's false-positive reductions. The foundation infra commit (`3ad8fc6e`) intentionally deferred the changeset downstream, and this PR had none — the `fixed` group cascades the bump to the sibling packages.

**Verification:** typecheck ✓ · full oxlint-plugin suite **7375 passed / 88 skipped** ✓ · lint ✓ · format ✓ · published `react-doctor --staged` on the diff → **100/100, no issues**.

**Follow-up (out of scope for this split):** `tanstack-start-no-navigate-in-render.ts` has an equivalent inline `isEventHandlerAttribute` that could be repointed to the new shared util in a dedup pass.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-nested-component-definition.ts [hunk 4]
```diff
@@ -16,14 +23,28 @@ export const noNestedComponentDefinition = defineRule({
...
+    // inlined into the parent's render (no separate fiber, no state to
+    // lose), so requiring JSX-render membership before reporting drops
+    // the inline-render-helper false positives.
+    const renderedJsxNames = new Set<string>();
 
     return {
+      JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
+        if (isNodeOfType(node.name, "JSXIdentifier") && isUppercaseName(node.name.name)) {
+          renderedJsxNames.add(node.name.name);
+        }
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-nested-component-definition.ts]: yes bug

**Review by rayhanadev — CHANGES_REQUESTED** [review]
devin/bugbot bugs

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-render-in-render.ts [hunk 5]
```diff
@@ -1,8 +1,93 @@
...
+// so calling it inline remounts nothing — the same render-prop carve-out
+// as the `props.renderX()` shape, just for the destructured spelling.
+// A locally-declared `renderRow` helper (kind "const"/"function" whose
+// source isn't a parameter) still carries the smell and stays flagged.
+const tracesToPropOrParameter = (
+  symbol: SymbolDescriptor | null,
+  scopes: ScopeAnalysis,
+): boolean => {
+  if (!symbol) return false;
+  if (symbol.kind === "parameter") return true;
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-render-in-render.ts]: Fixed in fef7dc44. The parameter carve-out is now scoped to **component** parameters via a shared `isComponentFunction` check, so a `render*` parameter of an ordinary nested helper (`function runRow(renderRow) { … }`) is flagged again — only a props object or a render prop destructured from a component parameter stays exempt. Regression test added: *still flags a render\* param of a nested non-component helper*.

File: packages/react-doctor/tests/regressions/prop-stack-barrier.test.ts [hunk 6]
```diff
@@ -71,6 +71,54 @@ export const Outer = ({ value }: { value: string }) => {
...
+    </div>
+  );
+};
+`,
+      },
+    });
+
+    const hits = await collectRuleHits(projectDir, "no-derived-useState");
+    expect(hits).toHaveLength(0);
+  });
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/react-doctor/tests/regressions/prop-stack-barrier.test.ts]: Correct — this case asserts the `no-derived-useState` re-seeded-draft-buffer FP fix, which is implemented in the **state-and-effects** PR (#990, where `no-derived-use-state.ts` gains `isReseededDraftBuffer`). It was misplaced on this architecture branch, which doesn't touch that rule (see the changeset scope), so it failed here. Removed in fef7dc44; the case rides with #990 alongside the rule change it depends on.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/design/no-outline-none.ts [hunk 7]
```diff
@@ -38,10 +60,13 @@ export const noOutlineNone = defineRule({
...
       if (!hasOutlineNone || !outlineProperty) return;
 
-      const hasCustomFocusRing = expression.properties?.some((property: EsTreeNode) => {
+      const hasInlineBoxShadowRing = expression.properties?.some((property: EsTreeNode) => {
         const key = getStylePropertyKey(property);
         return key === "boxShadow";
       });
+      const className = node.parent ? getStringFromClassNameAttr(node.parent) : null;
+      const hasClassNameFocusRing = Boolean(className && FOCUS_RING_CLASS_PATTERN.test(className));
+      const hasCustomFocusRing = hasInlineBoxShadowRing || hasClassNameFocusRing;
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/design/no-outline-none.ts:132]: Fixed in fef7dc44. `FOCUS_RING_CLASS_PATTERN` now ends with `(?!-(?:none|0)\b)`, so `focus:outline-none` / `focus:shadow-none` / `focus:ring-0` (removal utilities) no longer count as a replacement ring, while real rings like `focus:shadow-outline` / `focus-visible:ring-2` still do. Regression tests added for both directions.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-render-in-render.ts [hunk 8]
```diff
@@ -18,11 +103,16 @@ export const noRenderInRender = defineRule({
...
+        if (tracesToPropOrParameter(context.scopes.symbolFor(expression.callee), context.scopes)) {
+          return;
+        }
         calleeName = expression.callee.name;
       } else if (
         isNodeOfType(expression.callee, "MemberExpression") &&
         isNodeOfType(expression.callee.property, "Identifier")
       ) {
+        if (isRenderPropReceiver(expression.callee.object)) return;
+        if (isStableMethodReceiver(expression.callee.object)) return;
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-render-in-render.ts]: Fixed in fef7dc44. The inline member-call receiver now runs through `rootsInProps`, so `props.slots.renderItem()` is exempt for the same reason its destructured form (`const { renderItem } = props.slots`) already was. The old `isRenderPropReceiver` helper was removed in favor of `rootsInProps`. Regression test added.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-nested-component-definition.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-nested-component-definition.ts]: Fixed in fef7dc44. Rendered-JSX membership is now scoped to the candidate's own enclosing component via `isAstDescendant` (each `<Name/>` is kept with its node), so a sibling parent's `<Inner/>` no longer flags a same-named call-only helper in another parent. Regression test added: *does not leak a sibling's <Inner/> onto a same-named call-only helper*.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/rendering-svg-precision.ts [hunk 9]
```diff
@@ -4,10 +4,23 @@ import type { RuleContext } from "../../utils/rule-context.js";
...
+const SVG_PATH_HIGH_PRECISION_PATTERN = /\d+\.\d{4,}/g;
 
 const SVG_PATH_ATTRIBUTES = new Set(["d", "points", "transform"]);
 
+// Materiality gate: a single over-precise coordinate (even repeated) in a
+// one-off hand-written glyph saves only a handful of bytes once — not a
+// download cost worth a diagnostic. Real machine-exported / wasteful paths
+// carry many DISTINCT over-precise coordinates. Require at least this many
+// distinct over-precise tokens before reporting.
+const MIN_DISTINCT_OVERPRECISE_TOKENS = 2;
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/rendering-svg-precision.ts]: Fixed in fef7dc44. Moved to `packages/oxlint-plugin-react-doctor/src/plugin/constants/thresholds.ts` as `MIN_DISTINCT_OVERPRECISE_SVG_TOKENS` (with the explanatory comment) and imported it into the rule.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/no-prevent-default.ts [hunk 10]
```diff
@@ -64,6 +64,37 @@ const containsPreventDefaultCall = (node: EsTreeNode): boolean => {
...
+      isNodeOfType(callee, "MemberExpression") &&
+      isNodeOfType(callee.property, "Identifier") &&
+      NON_NAVIGATING_EVENT_METHOD_NAMES.has(callee.property.name)
+    ) {
+      return;
+    }
+    didFindSideEffectCall = true;
+  });
+  return didFindSideEffectCall;
+};
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/no-prevent-default.ts]: This FP-bias is intentional. The anchor message here is specifically *“nothing navigates”* — a dead link. A handler that also calls `setOpen(true)` / `track()` / a router push IS doing something on click, so *“nothing navigates”* would be false and the report a false positive (a link-styled menu/dropdown/modal toggle is a common, valid pattern). The distinct a11y concern — an `<a>` used as a button should be a `<button>` — is a different rule's domain, not this one's. So exempting any non-event-method call after `preventDefault()` is deliberate: we'd rather miss a genuinely dead `<a onClick={e => { e.preventDefault(); setOpen(true) }}>` than flag every interactive link. Leaving as-is.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-render-in-render.ts [hunk 11]
```diff
@@ -1,8 +1,89 @@
...
+    ) {
+      return true;
+    }
+    current = current.object;
+  }
+  if (isNodeOfType(current, "Identifier")) {
+    if (current.name === "props") return true;
+    return isComponentParameterSymbol(scopes.symbolFor(current));
+  }
+  return false;
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/architecture/no-render-in-render.ts:72]: Fixed in db7841f4. The `props` carve-out no longer matches by name — both `rootsInProps` and `tracesToPropOrParameter` now resolve the root identifier through scope via the shared `isComponentParameterSymbol` (a parameter of a React component). So a local `const props = { renderRow }` is no longer treated as the component's props bag, and `props.renderRow()` on it is flagged again. Regression test added: *still flags a render\* call on a local object named props*.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/no-polymorphic-children.ts [hunk 12]
```diff
@@ -1,9 +1,32 @@
...
+// True only when the `children` operand resolves to the enclosing
+// component's `props.children` — a destructured prop binding
+// (`({ children }) => …`, a `parameter` symbol) or a `props.children`
+// member access where `props` is the component's parameter. A local
+// variable or data field that happens to be named `children`
+// (`const { children } = node`) is not a polymorphic-children smell.
+const resolvesToPropsChildren = (operand: EsTreeNode, scopes: ScopeAnalysis): boolean => {
+  if (isNodeOfType(operand, "Identifier") && operand.name === "children") {
+    return scopes.symbolFor(operand)?.kind === "parameter";
+  }
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/no-polymorphic-children.ts:51]: Fixed in db7841f4. `resolvesToPropsChildren` now also flags a body alias of the props (`const { children } = props` / `const { children } = this.props`), which is the same smell as `({ children })` / `typeof props.children`. It stays component-aware, so the protected FP — `const { children } = node` where `node` is a parameter of a **non-component** tree walker — remains quiet. Regression test added: *flags `typeof children` when children is aliased from props in the body*.

**Review by rayhanadev — APPROVED** [review]
